src/main/java/com/dumbdogdiner/stickyapi/common/util/TextUtil.java

```diff
@@ -4,63 +4,143 @@
+ */
+@UtilityClass
 public class TextUtil {
-    static HashMap<Character, Integer> characterWidths = new HashMap<>() {
-        {
-            put('!', 1);
-            put(',', 1);
-            put('\'', 1);
-            put('.', 1);
-            put(':', 1);
-            put(';', 1);
-            put('i', 1);
-            put('|', 1);
-            put('!', 1);
-
-            put('`', 2);
-            put('l', 2);
-
-            put(' ', 3);
-            put('(', 3);
-            put(')', 3);
-            put('*', 3);
-            put('I', 3);
-            put('[', 3);
-            put(']', 3);
-            put('t', 3);
-            put('{', 3);
-            put('}', 3);
-
-            put('<', 4);
-            put('>', 4);
-            put('f', 4);
-            put('k', 4);
-
-            put('@', 6);
-            put('~', 6);
+    /** Offset for bold (in each direction) */
+    public static final float BOLD_OFFSET = 0.5f;
+    /** Offset for shadows (in each direction) */
+    public static final float SHADOW_OFFSET = 0.5f;
+
+    /** Number of pixels per line of a book */
+    public static final int PIXELS_PER_BOOK_LINE = 113;
+    /** Number of half-pixels per book line */
+    public static final int HALF_PIXELS_PER_BOOK_LINE = PIXELS_PER_BOOK_LINE * 2;
+
+    /** Number of lines in a page of a book */
+    public static final int LINES_PER_PAGE = 14;
+    /** Number of pages in a book */
+    public static final int PAGES_PER_BOOK = 50;
+
+    /** Number of pixels per line of a sign */
+    // TODO verify via minecraft source code
+    public static final int PIXELS_PER_SIGN_LINE = 96;
+    /** Number of lines per sign */
+    public static final int LINES_PER_SIGN = 4;
+
+
+    private static final Pattern SPLIT_PATTERN = Pattern.compile("(?<=[ \n])|(?=[ \n])");
+
+    private static final HashMap<Character, Integer> widths = new HashMap<>();
+
+    /**
+     * Internal class to parse the width data json file
+     */
+    @Data
+    private static class WidthEntry {
+        private @NotNull String uid;
+        private @Nullable Integer id;
+        private @NotNull String val;
+        private int width;
+    }
+
+    static {
+        Gson gson = new Gson();
+        try (InputStream input = ClassLoader.getSystemResource("generated/mojangles_width_data.json").openStream()) {
+            WidthEntry[] entries = gson.fromJson(new InputStreamReader(input), WidthEntry[].class);
+            for (WidthEntry entry : entries) {
+                if (entry.getId() == null) entry.setId(0);
+                widths.put((char) (entry.getId() != null ? entry.getId() : '\0'), entry.width);
+            }
+        } catch (Exception e) {
+            StickyAPI.getLogger().severe(e.getMessage());
+            // fallback to a minimal , in case anything fails
+            for (char c = 32; c <= 126; ++c) {
+                widths.put(c, 12);
+            }
+            "\0 ".chars().forEach(c -> widths.put((char) c, 2));
```

Comment:
This is really, really unfriendly code.

Reply:
Not my code, I believe it is @Prouser123

Reply:
CC: @SkyezerFox @aakatz3 
> Not my code, I believe it is @Prouser123

@spazzylemons wrote this in 3e4e10d6dc74c7c5886e13de96de03b919d5d51f as a fallback in case the JSON data could not be loaded for whatever reason.